### PR TITLE
Setup whitelist as OVM precompile

### PR DESCRIPTION
**Pull request by ben-chain**
## Description
This PR sets up the previously created `DeployerWhitelist` to be an OVM precompile, and fixes the context before it is called, so that it may use OVM storage.

## Usage
- The whitelist is associated with OVM address `0x4200000000000000000000000000000000000002`.
- The whitelist now uses `function initialize(address _owner, bool _allowArbitraryDeployment)` in place of a constructor.

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: packages/contracts/contracts/optimistic-ethereum/ovm/precompiles/DeployerWhitelist.sol [hunk 1]
```diff
@@ -89,9 +103,54 @@ contract DeployerWhitelist {
...
         external
-        view
         returns(bool)
     {
-        return allowArbitraryDeployment || whitelistedDeployers[_deployerAddress];
+        bool allowArbitraryDeployment = uint(doOvmSLOAD(ALLOW_ARBITRARY_DEPLOYMENT)) == 1;
+        bool isWhitelistedDeployer = uint(doOvmSLOAD(bytes32(bytes20(_deployerAddress)))) == 1;
+        bool isInitialized = uint(doOvmSLOAD(INITIALIZED_KEY)) != 0;
+        
+        return allowArbitraryDeployment || isWhitelistedDeployer || !isInitialized;
```
Comment (karlfloersch) [inline packages/contracts/contracts/optimistic-ethereum/ovm/precompiles/DeployerWhitelist.sol:112]: I feel slight regret that we won't eagerly exit this when we find that `allowArbitraryDeployment == true`, but that's totally fine for our initial version.

**Review by karlfloersch — APPROVED** [review]
This looks great! There are things that we should totally clean up but considering we're doing a comprehensive overhaul of this code I think we can very easily clean it up later.

File: packages/contracts/contracts/optimistic-ethereum/ovm/precompiles/DeployerWhitelist.sol [hunk 2]
```diff
@@ -5,24 +5,36 @@ pragma experimental ABIEncoderV2;
...
-    mapping(address=>bool) public whitelistedDeployers;
-    address public owner;
-    bool public allowArbitraryDeployment;
+    // mapping(address=>bool) public whitelistedDeployers;
+    bytes32 constant INITIALIZED_KEY = 0x0000000000000000000000000000000000000000000000000000000000000010;
+    bytes32 constant OWNER_KEY = 0x0000000000000000000000000000000000000000000000000000000000000011;
+    bytes32 constant ALLOW_ARBITRARY_DEPLOYMENT = 0x0000000000000000000000000000000000000000000000000000000000000012;
 
-    constructor(address _owner, bool _allowArbitraryDeployment)
+    constructor (address _owner, bool _allowArbitraryDeployment) public {}
```
Comment (karlfloersch) [inline packages/contracts/contracts/optimistic-ethereum/ovm/precompiles/DeployerWhitelist.sol:13]: Soon we need to remove this 😭

**Review by kfichter — APPROVED** [review]
LGTM 💵

**Comment by kfichter** [comment]
Obviously lots that could be cleaned up in the future, but this looks fine for now and can be simplified when we move to contracts v2.

**Pull request merged**